### PR TITLE
Make Dtensor have consistent cache key in compile

### PR DESCRIPTION
**Pull request by tugsbayasgalan**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #173526

Authored with Claude. Previously we were pickling the storage address which were causing different cache key on different processes. 

We should use _reduce_override protocol rather than dispatch_table protocol that AOTAutogradCache uses because the latter would require us to enumerate all subclass types manually. 

@diff-train-skip-merge this change was reverted internally

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @kadeng @chauhang @amjames @Lucaskabela @jataylo

### CODE REVIEW

File: torch/_functorch/_aot_autograd/autograd_cache.py [hunk 1]
```diff
@@ -440,6 +440,45 @@ def __init__(self, gm: torch.fx.GraphModule):
...
+        # Recursively handle inner tensors for nested subclasses
+        inner_metadata = {}
+        if is_traceable_wrapper_subclass(tensor):
+            inner_tensors, _ = tensor.__tensor_flatten__()
+            for name in inner_tensors:
+                inner_tensor = getattr(tensor, name)
+                if is_traceable_wrapper_subclass(inner_tensor):
+                    inner_metadata[name] = self._reduce_tensor_subclass(inner_tensor)
+
+        return (_ident, (metadata, repr(tensor), inner_metadata))
```
Comment (bdhirsh) [inline torch/_functorch/_aot_autograd/autograd_cache.py]: cc @bobrenjc93 (in case you're interested in reviewing AOTAutograd caching PRs!)

File: torch/_functorch/_aot_autograd/autograd_cache.py [hunk 2]
```diff
@@ -440,6 +440,45 @@ def __init__(self, gm: torch.fx.GraphModule):
             }
         )
 
+    # pyrefly: ignore [bad-override]
+    def reducer_override(self, obj: Any) -> Any:
```
Comment (bdhirsh) [inline torch/_functorch/_aot_autograd/autograd_cache.py:526]: I don't think I understand the override mechanism here. My understanding is roughly:

(1) the AOTAutograd cache needs to compute a cache on both the dynamo graph, **and** the dynamo example inputs (which may be DTensors)

(2) The code that gets a cache key for the example inputs ends up callin `__reduce_ex__` for each input, and for DTensor this does something non-deterministic (?)

Is this right? I think I'd want to know why e.g. storage addresses (or something else non-deterministic) show up in the DTensor __reduce_ex__ but not for plain tensor inputs. And if so, maybe we should implement our own better __reduce_ex__ for DTensor instead of adding this override mechanism?

Comment (tugsbayasgalan) [reply torch/_functorch/_aot_autograd/autograd_cache.py:526]: AOTAutograd cache has a default dispatch_table mechanism that directly dispatches plain tensor (https://github.com/pytorch/pytorch/blob/cd3825d33af02289b73f15bd7ad1dd0d3a1ac39e/torch/_functorch/_aot_autograd/autograd_cache.py#L502) as a result we never go through `__reduce_ex__` for plain tensors which has the storage issue. I think alternative way to implement is to add a similar dispatch mechanism for DTensor only but this is not good because we would have to do the same for all subclasses we care about. Hence, I added `reduce_override` handler to catch all.  If you think this is not good for performance reasons, we can go with `dispatch_table` approach.

File: test/dynamo/test_aot_autograd_cache.py [hunk 3]
```diff
@@ -2479,6 +2480,138 @@ def fn(x):
...
+            results[1],
+            "Nested tensor subclass cache keys should be identical across processes",
+        )
+
+    @unittest.skipIf(not torch.distributed.is_available(), "requires distributed")
+    def test_dtensor_different_process_cache_key(self):
+        """
+        Test that DTensor cache keys are consistent across different processes.
+
+        This is a critical test for warm start cache hits. The cache key must be
```
Comment (bdhirsh) [inline test/dynamo/test_aot_autograd_cache.py:3549]: the test requiring multiprocessing is a bit strange to me 🤔 . Does this test fail properly without your fix? If you look at most of the other AOTAutograd cache tests, I think they mostly just reset dynamo in the same process and run compile twice to confirm cache miss / cache hit behavior.

File: torch/_functorch/_aot_autograd/autograd_cache.py [hunk 4]
```diff
@@ -440,6 +440,45 @@ def __init__(self, gm: torch.fx.GraphModule):
...
+        """
+        from torch._inductor.codecache import extract_tensor_metadata_for_cache_key
+        from torch.utils._python_dispatch import is_traceable_wrapper_subclass
+
+        metadata = extract_tensor_metadata_for_cache_key(tensor)
+
+        # Recursively handle inner tensors for nested subclasses
+        inner_metadata = {}
+        if is_traceable_wrapper_subclass(tensor):
+            inner_tensors, _ = tensor.__tensor_flatten__()
```
Comment (bdhirsh) [inline torch/_functorch/_aot_autograd/autograd_cache.py]: this ignores the metadata returned by __tensor_flatten__ (the second output), right?  I think that means that we will ignore all tensor subclass metadata when computing the cache key

Comment (bdhirsh) [reply torch/_functorch/_aot_autograd/autograd_cache.py]: So a good negative test to add would be that if you run a small fn twice, where the input is a DTensor with Replicat() the first time and a DTensor with Shard(0) the second time, we should cache miss both times.

File: test/dynamo/test_aot_autograd_cache.py (hunk 3 above)
Comment (tugsbayasgalan) [reply test/dynamo/test_aot_autograd_cache.py:3549]: Yes it does fail without my fix. I can follow the other test patterns if needed. This is mainly me trying to replicate the real user behaviour lol. Previously when we had an issue with hashing due to dynamo source, none of these tests catch the issue so i thought we don't actually test different processes lol.

Comment (bdhirsh) [reply test/dynamo/test_aot_autograd_cache.py:3549]: sg. If we're able to repro an error without multiprocessing that would probably be better. but if it only repros with multiprocessing then sounds fine!

File: torch/_functorch/_aot_autograd/autograd_cache.py (hunk 4 above)
Comment (tugsbayasgalan) [reply torch/_functorch/_aot_autograd/autograd_cache.py]: This works because we call repr on the whole tensor later on. But yeah, good point. We should probably do repr(metadata) instead.

File: torch/_functorch/_aot_autograd/autograd_cache.py [hunk 5]
```diff
@@ -503,6 +503,45 @@ def __init__(self, gm: torch.fx.GraphModule):
...
+        inner_metadata = {}
+        subclass_metadata = None
+        if is_traceable_wrapper_subclass(tensor):
+            inner_tensors, subclass_metadata = tensor.__tensor_flatten__()
+            for name in inner_tensors:
+                inner_tensor = getattr(tensor, name)
+                if is_traceable_wrapper_subclass(inner_tensor):
+                    inner_metadata[name] = self._reduce_tensor_subclass(inner_tensor)
+
+        return (_ident, (metadata, repr(subclass_metadata), inner_metadata))
```
Comment (bdhirsh) [inline torch/_functorch/_aot_autograd/autograd_cache.py]: hmm i am a bit worried about relying on users implemented `repr()` on their subclass metadata in order to get reasonable caching behavior. Today, dyamo relies on subclasses implementing `==` in a reasonable way to get guards on metadata working.

We could use `hash()`, since python objects that implement custom equality are on the hook for implemented a custom hash anyway? Although I'm not sure if e.g. DeviceMesh has a reasonable hash implementation today. For example, if DeviceMesh hash includes hashing on ProcessGroup info that can change run-to-run you will end up with unnecessary cache misses.

I think my vote would be to see if we can use hash(), and check if DeviceMesh.__hash__ does something reasonable that doesn't cause equivalent device meshes to get a different hash from run to run. And if it does, maybe we can check with distributed folks to see how reasonable it would be to change this (cc @fegin ?)

File: test/dynamo/test_aot_autograd_cache.py [hunk 6]
```diff
@@ -48,6 +51,20 @@
 )
 
 
+@contextlib.contextmanager
```
Comment (aorenste) [inline test/dynamo/test_aot_autograd_cache.py:79]: This seems generally useful - maybe move to torch/testing/_utils.py? (There are a ton of calls to `init_process_group("fake", ...)` in various tests.

File: torch/_functorch/_aot_autograd/autograd_cache.py [hunk 7]
```diff
@@ -503,6 +503,72 @@ def __init__(self, gm: torch.fx.GraphModule):
...
+        #
+        # DTensor's __tensor_flatten__ returns (DTensorSpec, requires_grad)
+        if (
+            isinstance(subclass_metadata, tuple)
+            and len(subclass_metadata) >= 1
+            and isinstance(subclass_metadata[0], DTensorSpec)
+        ):
+            return repr(subclass_metadata)
+
+        return hash(subclass_metadata)
```
Comment (aorenste) [inline torch/_functorch/_aot_autograd/autograd_cache.py]: hash is definitely NOT cross-process stable because the seed changes on each python invocation... (the comment says "may not")

File: torch/_functorch/_aot_autograd/autograd_cache.py [hunk 8]
```diff
@@ -503,6 +503,72 @@ def __init__(self, gm: torch.fx.GraphModule):
...
+        # stable repr() implementations, we use repr() here for DTensor specifically.
+        # Other subclasses fall back to hash() which may not be cross-process stable.
+        #
+        # DTensor's __tensor_flatten__ returns (DTensorSpec, requires_grad)
+        if (
+            isinstance(subclass_metadata, tuple)
+            and len(subclass_metadata) >= 1
+            and isinstance(subclass_metadata[0], DTensorSpec)
+        ):
+            return repr(subclass_metadata)
```
Comment (aorenste) [inline torch/_functorch/_aot_autograd/autograd_cache.py]: relying on repr() being stable + explicit enough for a hash seems dangerous (like it seems weird that I can change repr on DeviceMesh and suddenly DTensor caching stops working even though there's no obvious link between the two). Should we define something like "`__tensor_hash__`" on tensor subclasses which defaults to a reasonable stable hash but can be overridden on the DTensor itself?

File: torch/_functorch/_aot_autograd/autograd_cache.py (hunk 7 above)
Comment (tugsbayasgalan) [reply torch/_functorch/_aot_autograd/autograd_cache.py]: Yeah... I also don't want to rely on user defined repr() implementation tho. I was mainly going off of (https://github.com/pytorch/pytorch/pull/173526#discussion_r2747942182).

File: torch/_functorch/_aot_autograd/autograd_cache.py (hunk 8 above)
Comment (tugsbayasgalan) [reply torch/_functorch/_aot_autograd/autograd_cache.py]: Yeah that could work. Let me try that. I am thinking we would have if/else where we do stable and slow hashing like blake2 or sth for compile AND just default hash for eager. The reason is we don't want to hurt perf for eager. Wdyt?

File: torch/_functorch/_aot_autograd/autograd_cache.py [hunk 9]
```diff
@@ -513,6 +519,98 @@ def __init__(self, gm: torch.fx.GraphModule) -> None:
...
+        for name in inner_tensor_names:
+            inner_tensor = getattr(tensor, name)
+            inner_hashes[name] = self._get_stable_hash(inner_tensor)
+
+        cache_data = repr(
+            (
+                tensor.shape,
+                tensor.dtype,
+                tensor.device,
+                tensor.requires_grad,
```
Comment (aorenste) [inline torch/_functorch/_aot_autograd/autograd_cache.py:619]: Doesn't this have to be part of the metadata returned by `__tensor_flatten__`? Otherwise how would unflatten be able to reconstruct the subclass properly?

File: torch/_functorch/_aot_autograd/autograd_cache.py [hunk 10]
```diff
@@ -513,6 +519,98 @@ def __init__(self, gm: torch.fx.GraphModule) -> None:
...
+        """
+        inner_tensor_names, subclass_metadata = tensor.__tensor_flatten__()  # type: ignore[attr-defined]
+
+        # Recursively get hashes of inner tensors
+        inner_hashes = {}
+        for name in inner_tensor_names:
+            inner_tensor = getattr(tensor, name)
+            inner_hashes[name] = self._get_stable_hash(inner_tensor)
+
+        cache_data = repr(
```
Comment (aorenste) [inline torch/_functorch/_aot_autograd/autograd_cache.py]: Do we not want to pickle this because we're worried about unpicklable metadata? Now that a tensor can provide its own _stable_hash_for_caching I would think it would be "fine".

File: torch/_functorch/_aot_autograd/autograd_cache.py [hunk 11]
```diff
@@ -513,6 +519,98 @@ def __init__(self, gm: torch.fx.GraphModule) -> None:
...
+        """
+        Reduce a tensor subclass to a stable key for caching.
+        """
+        from torch.utils._python_dispatch import is_traceable_wrapper_subclass
+
+        if not is_traceable_wrapper_subclass(tensor):
+            return self._reduce_tensor(tensor)
+
+        if hasattr(tensor, "_stable_hash_for_caching"):
+            return (_ident, (tensor._stable_hash_for_caching(),))
```
Comment (aorenste) [inline torch/_functorch/_aot_autograd/autograd_cache.py]: Maybe check _stable_hash_for_caching first - because we don't need `__tensor_flatten__` (which traceable implies) if it has that defined.

File: torch/_functorch/_aot_autograd/autograd_cache.py [hunk 12]
```diff
@@ -513,6 +519,98 @@ def __init__(self, gm: torch.fx.GraphModule) -> None:
             }
         )
 
+    # pyrefly: ignore [bad-override]
+    def reducer_override(self, obj: Any) -> Any:
```
Comment (aorenste) [inline torch/_functorch/_aot_autograd/autograd_cache.py:526]: nit: 
```suggestion
    def reducer_override(self, obj: object) -> tuple[Callable[..., Any], tuple[Any, ...]]:
```

File: torch/_functorch/_aot_autograd/autograd_cache.py [hunk 13]
```diff
@@ -513,6 +519,98 @@ def __init__(self, gm: torch.fx.GraphModule) -> None:
...
+        Override to handle tensor subclasses (like DTensor) that aren't caught
+        by the dispatch_table's exact type matching.
+
+        The dispatch_table only matches exact types, so subclasses like DTensor
+        fall through to the default __reduce_ex__ which includes non-deterministic
+        storage addresses. This method catches those cases using isinstance checks.
+        """
+        # Handle tensor subclasses that aren't exactly torch.Tensor
+        # dispatch_table already handles torch.Tensor exactly
+        if isinstance(obj, torch.Tensor) and type(obj) is not torch.Tensor:
```
Comment (aorenste) [inline torch/_functorch/_aot_autograd/autograd_cache.py:539]: Will the "is not torch.Tensor" ever be false? (i.e. doesn't the dispatch table kick in first so we should never see a vanilla Tensor?)

File: torch/_functorch/_aot_autograd/autograd_cache.py [hunk 14]
```diff
@@ -513,6 +519,98 @@ def __init__(self, gm: torch.fx.GraphModule) -> None:
...
+        from torch.utils._python_dispatch import is_traceable_wrapper_subclass
+
+        if not is_traceable_wrapper_subclass(tensor):
+            return self._reduce_tensor(tensor)
+
+        if hasattr(tensor, "_stable_hash_for_caching"):
+            return (_ident, (tensor._stable_hash_for_caching(),))
+
+        # Warn once about missing _stable_hash_for_caching
+        warn_once(
```
Comment (aorenste) [inline torch/_functorch/_aot_autograd/autograd_cache.py]: I don't know how i feel about this warning.  We warn about tensor subclasses that implement (tensor_flatten and not _stable_hash_for_caching) but NOT about tensor subclasses that don't implement tensor_flatten at all?

File: torch/_functorch/_aot_autograd/autograd_cache.py [hunk 15]
```diff
@@ -513,6 +519,98 @@ def __init__(self, gm: torch.fx.GraphModule) -> None:
...
+    # generation by implementing _stable_hash_for_caching(self) -> str.
+    # This method should return a deterministic string that uniquely identifies
+    # the tensor's metadata for caching purposes. See DTensor for an example.
+    #
+    # We can't define a default method on subclasses because there is no abstract
+    # base subclass, and we don't want to pollute torch.Tensor. Instead, we provide
+    # a default implementation here that uses __tensor_flatten__ to recursively
+    # hash inner tensors and metadata.
+
+    def _reduce_tensor_subclass(self, tensor: torch.Tensor) -> Any:
```
Comment (aorenste) [inline torch/_functorch/_aot_autograd/autograd_cache.py]: Maybe we don't need this function? If reducer_override just called _get_stable_hash() directly and returned something like `(_ident, (_get_stable_hash(t),))`?

**Review by aorenste — APPROVED** [review]
A few comments but looks generally reasonable (and I'm on PTO next week)

File: torch/_functorch/_aot_autograd/autograd_cache.py (hunk 9 above)
Comment (tugsbayasgalan) [reply torch/_functorch/_aot_autograd/autograd_cache.py:619]: I think unflatten takes in shape parameter directly to construct a subclass. But yeah dtype and device are redundant here.

File: torch/_functorch/_aot_autograd/autograd_cache.py (hunk 10 above)
Comment (tugsbayasgalan) [reply torch/_functorch/_aot_autograd/autograd_cache.py]: Sure! I think we can pickle now

File: torch/_functorch/_aot_autograd/autograd_cache.py (hunk 13 above)
Comment (tugsbayasgalan) [reply torch/_functorch/_aot_autograd/autograd_cache.py:539]: yes because dispatcher table kicks in later than reducer_override function.

**Review by seemethere — APPROVED** [review]
Review automatically exported from Phabricator review in Meta.
